File: 19_remove_nth_node_from_end_of_list/removeNthFromEnd.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	right := head
	left := head

	for i := 0; i < n; i++ {
		right = right.Next
	}

	if right == nil {
		return head.Next
	}

	for right.Next != nil {
		right = right.Next
		left = left.Next

	}
	left.Next = left.Next.Next

	return head
}

File: 19_remove_nth_node_from_end_of_list/removeNthFromEnd.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        l = r = head

        for i in range(n):
            l = l.next

        if not l:
            return head.next

        while l.next:
            l = l.next
            r = r.next

        r.next = r.next.next
        return head
            
